### ISSUE TITLE
Skopeo packaged for conda-forge

### ISSUE THREAD

**Issue by epruesse** [containers/skopeo#643]
Just FYI: I've packaged skopeo for conda-forge: See https://github.com/conda-forge/skopeo-feedstock

It's just the `DISABLE_CGO=1` variant, and no OSX or Windows builds for now.

I also could not get the `0.1.35` release to work, so I stuck to the `0.1.32`. 

Build script: https://github.com/conda-forge/skopeo-feedstock/blob/master/recipe/build.sh
Recipe: https://github.com/conda-forge/skopeo-feedstock/blob/master/recipe/meta.yaml

**Comment by rhatdan** [comment]
Nice, Thanks.
Are you going to try to get it to work on OSX and Windows?
What issues were you having with 0.1.35?

How does this work?  Does Conda install on any distribution?  Are you building skopeo statically?

Any thoughts on buildah or podman?

**Comment by TomSweeneyRedHat** [comment]
Once the dust settles, it would be nice to document this in the README.md or some other appropriate location.

**Comment by epruesse** [comment]
To be honest, creating the package was need driven. I had been looking for something that can upload a docker archive to a registry without needing access to the docker daemon, and was VERY happy to find this project!

> Are you going to try to get it to work on OSX and Windows?

If I have time to spare. You know how that goes. Don't bank on it. It shouldn't be too hard, though. PRs on the repo listed above are welcome from anyone. Basically, remove `osx` or `linux` from the `skip` line and see what happens on the Azure check. I don't recall what the problem was on OSX, and I have little experience with windows.

> What issues were you having with 0.1.35?

It wouldn't built initially, and then I got a binary that segfaulted. Might have been me. I needed this fast, and didn't dig into it. This was my first Go package and I haven't yet had a chance to learn the language, so I ran into the usual first-timer problems (GOPATH, vendoring, go vs cgo, etc).

> How does this work? Does Conda install on any distribution?

Conda installs anywhere without root. Usually, you'd bootstrap by downloading and running the `miniconda3` installer, which will create a folder containing python and the `conda` installer and put the hooks into your shell config files. You could also start with the `continuumio/miniconda3` docker image, which is basically just that on a minimal system. You can then `conda install <package>` to install packages. It will download and install into your user account. 

One of the neat features of conda is that anaconda.org offers free hosting for custom distribution channels. Bioconda (for which I needed skopeo to use in our build infrastructure) and Conda-Forge are the biggest ones. To "tap" a channel by default, you call `conda config --add channels conda-forge`. Alternatively, you can do one-off using `conda install -c conda-forge skopeo`.

The second great feature is the isolated environments. By doing `conda create -n myenv skopeo` you create a new folder structure with initially just skopeo and dependencies (if it had any). The binaries and scripts get their paths rewritten as needed, everything else is hard-linked to minimize disk usage. Things in the environment can be accessed either via `conda activate myenv` to get access (rewrites PATH & co), or via full path to the installed tools (which works because Python/Perl/R/Lua and all DSOs have their internal paths rewritten).

Think of it as a package manager that folds in PyPi, CRAN, CPAN & Co but will also install binaries such as Go packages or other compiled things without needing root.

The third great part, for science at least, is that packages never go away. You can always install `conda install skopeo=0.1.32`. Helps reproducibility greatly. The package resolution isn't ad-hoc as with `pip`, but uses a full SAT solver. That means that it is often slow, much slower than say `apk install`, but then the list of packages it has to consider is huge and it does try to find an optimal solution for often complex mixtures of packages, with no manual "releases" involved. Combined with the environments, that makes it possible to have a current tool depend on a specific old version of something else. Science software can be moving targets, and since people are usually still learning, code quality is often low and API stability something people haven't heard of. 

> Are you building skopeo statically?

I'm building `make binary-local DISABLE_CGO=1`. So, no, not statically. But also not linking to `btrfs` etc. System libraries are tough with conda, because of the non-root nature. A lot of it is possible (we have Xorg after all), but I'm not sure how much sense it makes. I suppose `gpgme` would be something to package - not sure what the other libs are needed for. The binary produced without CGO was working fine for me (but my task was only: take that `docker save`d file and `docker push` it without having a daemon). 

Conda packages don't have to build statically, either. I have three sections of dependencies - the `host` lists the build tools on the build host, `build` lists the libs on the target architecture, `run` the libs and tools needed during runtime. If I list `gpgme` as `build` and `run` dependency, then I can use `pin_compatible(gpgme)` to automatically constrain the runtime versions to within the same minor, at least the patch level used at build time. Often, upstream packages will actually export this, so for e.g. `zlib`, I don't have to say anything, if it's in `build`, it will become a `run` time dependency pinned as specified by the `zlib` package itself.

Anway. I actually use conda to get software into docker containers. That's why I needed to package `skopeo` - so that I can just add it to our `requirements` (vendors) file and have the right version be pulled into the docker container automatically.

**Comment by epruesse** [comment]
Sorry about the long post...

> Any thoughts on buildah or podman?

Well, PRs are welcome, the process is actually quite simple. I'd be happy to help, but I don't anticipate having much time in the near future to spend on this (work related).

**Comment by epruesse** [comment]
Looks like my 0.1.35 issue might have been local:

Conda-Forge's update bot picked up that there is a newer release and it seems to build fine: 
- 0.1.33 https://github.com/conda-forge/skopeo-feedstock/pull/2
- 0.1.35 https://github.com/conda-forge/skopeo-feedstock/pull/2

I'll merge them. Might need some better testing than just `skopeo --help`.

**Comment by epruesse** [comment]
OSX package done

**Comment by epruesse** [comment]
0.1.36 does not build due to problem in buildah

**Comment by TomSweeneyRedHat** [comment]
@eprusse can you share the error(s)?

**Comment by epruesse** [comment]
- Opened issue here: https://github.com/containers/buildah/issues/1600
- Conda-Forge build here: https://github.com/conda-forge/skopeo-feedstock/pull/4

**Comment by epruesse** [comment]
It's the use of `memfd.h`, which is missing in older Linux. Conda-Forge builds on as old as possible GLIBC so that binaries will work everywhere. Looks like `memfd.h` is "newish".

**Comment by rhatdan** [comment]
Is this still blocked.  Does skopeo still require memfs.h?  Since this Issue is so old, I am going to assume that it is been fixed or is no longer desired, so I will close. Reopen if I am mistaken.

**Comment by epruesse** [comment]
@rhatdan - @mbargull is working on getting all `containers` tools packaged. He'll know better than I, but I think the issue has been resolved in the mean time.

**Comment by mbargull** [comment]
Yes, we still build on a CentOS 6 container and with some patches it works fine for now. I'm also more comfortable with maintaining patches for that myself (until we finally update to CentOS 7) than to offload more maintenance burden on upstream ;).

----
As for the other packages:
They should be built/released this week, I hope (only review pending, otherwise ready).
I've opened an issue over at https://github.com/containers/common/issues/346 to describe what I had to do to make everything work for Conda installations. I look forward to that discussion :).
